### PR TITLE
Enhance Crypto

### PR DESCRIPTION
**Pull request by milleruntime**
* Make crypto configurable per table
* Create separate crypto config for table and wal
* Create separate config for encrypt and decrypt
* Move init method on CryptoService interface into encrypt and decrypt
* Make CryptoUtils only parse service name from decryption
bytes in order to get decrypter
* Add multiple decrypters to allow reading different files
* Replace AESCryptoService with AESTableCryptoService and
AESWALCryptoService
* Add new tests for changes

### CODE REVIEW

**Comment by milleruntime** [comment]
I think these changes are good to go for review. I think there is still more follow on work but they are in a good place and I don't want this PR getting any bigger. One improvement I want to do as a follow on is to cache the decrypters so we don't have to call init every time we read a file.

File: core/src/main/java/org/apache/accumulo/core/crypto/CryptoServiceFactory.java [hunk 1]
```diff
@@ -56,11 +96,50 @@ public static CryptoService newInstance(AccumuloConfiguration conf, ClassloaderT
...
-    return newCryptoService;
+    var encrypter = newCryptoService.getEncrypter();
+    encrypter.init(initParams);
+    return encrypter;
   }
 
-  public static CryptoService newDefaultInstance() {
-    return newInstance(DefaultConfiguration.getInstance(), ClassloaderType.JAVA);
+  public static List<CryptoService> getDecrypters(AccumuloConfiguration conf, ClassloaderType ct) {
+    String[] classes = conf.get(TABLE_CRYPTO_DECRYPT_SERVICES).split(",");
```
Comment (PircDef) [inline core/src/main/java/org/apache/accumulo/core/crypto/CryptoServiceFactory.java:105]: Should this also load the TABLE_CRYPTO_ENCRYPT_SERVICE decrypter under the expectation that the service currently configured to encrypt will also be needed to decrypt? If not, should we add a note to the Property.java making it clear that the service much be set in both locations?

File: core/src/main/java/org/apache/accumulo/core/crypto/CryptoUtils.java [hunk 2]
```diff
@@ -84,4 +84,65 @@ public static void writeParams(byte[] decryptionParams, DataOutputStream out) th
...
+    CryptoService cs = null;
+    // check if service in file matches what was loaded from config
+    for (CryptoService d : decrypters) {
+      if (d.getClass().getName().equals(nameInFile)) {
+        cs = d;
+        break;
+      }
+    }
+    if (cs == null)
+      throw new CryptoException("Unknown crypto class found in " + scope + " file: " + nameInFile);
```
Comment (PircDef) [inline core/src/main/java/org/apache/accumulo/core/crypto/CryptoUtils.java:115]: This is the only place I see the scope being used. With the separation of WAL and RFile crypto services, do we still need this?

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/crypto/CryptoUtils.java:115]: I am not sure but I thought we still need a way to tell the CryptoService which type of file, since a user can still implement one to be used for both.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 3]
```diff
@@ -166,21 +166,6 @@
...
-  @Sensitive
-  INSTANCE_CRYPTO_SENSITIVE_PREFIX("instance.crypto.opts.sensitive.", null, PropertyType.PREFIX,
-      "Sensitive properties related to on-disk file encryption."),
-  @Experimental
-  INSTANCE_CRYPTO_SERVICE("instance.crypto.service",
-      "org.apache.accumulo.core.spi.crypto.NoCryptoService", PropertyType.CLASSNAME,
-      "The class which executes on-disk file encryption. The default does nothing. To enable "
-          + "encryption, replace this classname with an implementation of the"
-          + "org.apache.accumulo.core.spi.crypto.CryptoService interface."),
-
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:183]: The 2.0 crypto service concept was based on the idea that crypto policies should be centrally enforced.

Rather than replace the central crypto service with a multitude of separate services per table and WAL, I think you could extend the existing central service concept to support additional config for separate tables and WAL, and to change the implmentation to pass per-table and WAL context information into the central crypto service.

Have you considered this already?

Have you considered extending the 2.0 single-factory concept by supporting the ability to have a single crypto factory/module that can be dropped in and has manages the separate crypto settings based on separate per-table or WAL contexts, rather than requiring separately configured modules? I ask because I think there's value in ensuring that crypto policies are centrally enforced.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:183]: I did at first but it was easier for me to implement the division of table and WAL initially with them split up. I had problems at first since it was ambiguous in the Accumulo code what the `CryptoService` was doing. The tablet server for example initialized one at startup that could be used for either. I had separate interfaces as well but this led to a lot of duplicate code. Now that I split up the encrypt and decrypt across our mess of internal file builders, it may be easier to go back and combine them into a central service.

File: core/src/main/java/org/apache/accumulo/core/crypto/CryptoServiceFactory.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/crypto/CryptoServiceFactory.java:105]: That isn't a bad idea. I was going with the approach that if the decrypt service isn't in the config that we won't load it. Another option would be to make it flexible by adding the encrypt service to the list of decrypters if it isn't present.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 3 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:183]: One issue I had was trying to specify an inner class in the configuration. I am not sure why but this Java notation didn't seem to work:

`org.apache.accumulo.core.spi.crypto.CryptoService$AESGCMCryptoModule`

Any ideas if this can work with our classloading and/or configuration?

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:183]: I think the issue I had was with the way I was loading classes in a previous change. It seems like it should work now with these changes. The only issue is that it is a bit long to configure:
`tserver.walog.crypto.decrypt.service=org.apache.accumulo.core.spi.crypto.AESCryptoService$AESWALCryptoService`

**Comment by milleruntime** [comment]
@ctubbsii let me know what you think of the changes in 82dd8a5 before I take the time to change the names in the tests.

**Comment by milleruntime** [comment]
> @ctubbsii let me know what you think of the changes in [82dd8a5](https://github.com/apache/accumulo/commit/82dd8a546cc2fcf53ac0b4910299925ed99bf693) before I take the time to change the names in the tests.

I settled on `AESCryptoService` with inner classes of `Table` and `WAL`. So to configure encryption a user needs these properties:
<pre>
tserver.walog.crypto.encrypt.service=org.apache.accumulo.core.spi.crypto.AESCryptoService$WAL
tserver.walog.crypto.decrypt.service=org.apache.accumulo.core.spi.crypto.AESCryptoService$WAL
tserver.walog.crypto.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile
table.crypto.encrypt.service=org.apache.accumulo.core.spi.crypto.AESCryptoService$Table
table.crypto.decrypt.services=org.apache.accumulo.core.spi.crypto.AESCryptoService$Table
table.crypto.opts.key.uri=/home/mike/workspace/uno/CryptoTest-testkeyfile
</pre>

**Comment by ctubbsii** [comment]
> > @ctubbsii let me know what you think of the changes in [82dd8a5]

I still prefer the idea of keeping the crypto stuff contained in a single module. Something like:

```properties
instance.crypto.service=com.example.crypto.service.ExampleCryptoService
instance.crypto.opts.wal.SOMEKEY=SOMEVALUE
instance.crypto.opts.table.t1.SOMEKEY=SOMEVALUE
instance.crypto.opts.table.t2.SOMEKEY=SOMEVALUE
```

or (depending on implementation):

```properties
instance.crypto.service=com.example.crypto.service.ExampleCryptoService
instance.crypto.opts.wal.SOMEKEY=SOMEVALUE
; a per-table setting that works because the crypto service is a factory for encryptor/decryptor
; that is capable of making decisions based on per-table config
table.crypto.opts.SOMETABLEOPT=SOMEVALUE
```

With both of these, the main point is that there is only ever a single crypto service for the system. However, the functionality of that service is capable of providing crypto services for multiple tables and/or the WAL.

**Comment by milleruntime** [comment]
> instance.crypto.opts.table.t1.SOMEKEY=SOMEVALUE

I don't think this will work with the way Accumulo handles table properties differently. Also aren't `instance` properties used to determine the hash for a cluster? It would also be confusing because it breaks the delineations set by the prefix properties.

> With both of these, the main point is that there is only ever a single crypto service for the system. However, the functionality of that service is capable of providing crypto services for multiple tables and/or the WAL. Your change just relocates the implementation into inner classes, but they are still effectively separate classes that are configured separately, meaning separate authorities for managing the file-based crypto.

Yes but the crypto implementation for Table class won't work with WALs, due to the type of encryption. I was trying to keep them separate enough to make this clear that not all types of crypto will work with WALs vs Tables but keep them together enough to implement the same interfaces.

I believe what you want is what we have now, passing in the environment to the impl and letting it decided Table vs WAL. The problem is Accumulo needs to know whether the CryptoService is being used per table or tserver (WAL). Otherwise it gets very confusing looking at the code and trying to decipher the purpose of the CryptoService. Some of this can be derived from the code, for example within RFile or BCFile is obvious. But other parts are not, like the Tserver or FileUtilties. We can make sure to name variables based on use but the CryptoServiceFactory is a static utility.

**Comment by PircDef** [comment]
> With both of these, the main point is that there is only ever a single crypto service for the system. However, the functionality of that service is capable of providing crypto services for multiple tables and/or the WAL. Your change just relocates the implementation into inner classes, but they are still effectively separate classes that are configured separately, meaning separate authorities for managing the file-based crypto.

One of the goals for the crypto service in 2.0 was to make it more modular. Admins should be able to replace crypto services as necessary. Any new service needs to be capable of decrypting existing data. This means it needs to be able to decrypt any data encrypted by all prior services. As it stands now, all decrypters from all prior services would need to be part of the current crypto service. These changes remedy that by allowing prior crypto services to continue to provide their own decrypters. 

There is a case to be made for table specific crypto services. Allowing for table-level configuration enables data owners to control the crypto for their data in their table. Consider a case where someone wants to bulk import existing RFiles into a new table in another Accumulo instance. Lets assume they use a custom crypto service. With table-level crypto services, they would need to provide their RFiles and crypto service in order to read the data without requiring any changes to existing crypto services.

I think it is a good idea to explicitly specify the default crypto service used by Accumulo if a table is not specifically configured. This could allow for a single authority when appropriate. Would another option such as:

    instance.allow.table.crypto=false

be a reasonable compromise? This would allow admins to explicitly enable or disable table-level crypto services.

**Comment by ctubbsii** [comment]
> > instance.crypto.opts.table.t1.SOMEKEY=SOMEVALUE
> 
> I don't think this will work with the way Accumulo handles table properties differently. Also aren't `instance` properties used to determine the hash for a cluster? It would also be confusing because it breaks the delineations set by the prefix properties.

`general.*` would also work (and could be changed at runtime, unlike `instance.*`). The main point here is that these aren't characteristics of the table, but rather, characteristics of the crypto service module that inform it of how to behave in specific contexts.

> > With both of these, the main point is that there is only ever a single crypto service for the system. However, the functionality of that service is capable of providing crypto services for multiple tables and/or the WAL. Your change just relocates the implementation into inner classes, but they are still effectively separate classes that are configured separately, meaning separate authorities for managing the file-based crypto.
> 
> Yes but the crypto implementation for Table class won't work with WALs, due to the type of encryption.

Right, that's why the config is associated with the crypto service module, as opts, rather than tied to a table's characteristics. What opts work for which contexts, depends largely on the crypto service module's capabilities and is not solely an inherent characteristics of Accumulo objects (tables, WALs, etc.). Offloading this responsibility to the module

> I was trying to keep them separate enough to make this clear that not all types of crypto will work with WALs vs Tables but keep them together enough to implement the same interfaces.

That makes sense, but I don't think we need to have separate services conforming to the same interface for these. We can have a single service interface that provides services for both streaming encrypters/decrypters suitable for WAL, and for block encrypters/decrypters suitable for RFiles. If we really wanted to have them conform to the same interface, maybe that's okay, but instead of merely being colocated and configured separately, they could dangle off a single crypto service interface as sub-services, so you only have to configure the main crypto service module, and that acts as a factory for the sub-services that dangle from it.

> I believe what you want is what we have now, passing in the environment to the impl and letting it decided Table vs WAL.

I do lean towards the design principles that were put into the earlier effort, yes.

> The problem is Accumulo needs to know whether the CryptoService is being used per table or tserver (WAL). Otherwise it gets very confusing looking at the code and trying to decipher the purpose of the CryptoService. Some of this can be derived from the code, for example within RFile or BCFile is obvious. But other parts are not, like the Tserver or FileUtilties. We can make sure to name variables based on use but the CryptoServiceFactory is a static utility.

I think we can adapt the existing design by incorporating additional context information into the API, so the crypto service module can make decisions about the encrypters/decrypters to provide, based on the context it is given.

In general, what I'm arguing for is an API design that looks like:

```java
public interface SomeService {
  public SomeObject getSomeObject(Context ctx);
}
```

rather than one that looks like:

```java
public interface SomeService {
  public SomeObject getSomeObject();
}

public class SomeServiceForContext1 extend SomeService {
  public SomeObject getSomeObject() {};
}

public interface SomeServiceForContext2 extend SomeService {
  public SomeObject getSomeObject() {};
}
```

**Comment by ctubbsii** [comment]
> > With both of these, the main point is that there is only ever a single crypto service for the system. However, the functionality of that service is capable of providing crypto services for multiple tables and/or the WAL. Your change just relocates the implementation into inner classes, but they are still effectively separate classes that are configured separately, meaning separate authorities for managing the file-based crypto.
> 
> One of the goals for the crypto service in 2.0 was to make it more modular. Admins should be able to replace crypto services as necessary. Any new service needs to be capable of decrypting existing data. This means it needs to be able to decrypt any data encrypted by all prior services. As it stands now, all decrypters from all prior services would need to be part of the current crypto service. These changes remedy that by allowing prior crypto services to continue to provide their own decrypters.

A centralized crypto service module is not necessarily monolithic. It can be composed of modular components. But that granularity, if required, is now seen in the construction and configuration of a single drop-in service, rather than exposed as a multitude of configuration options in Accumulo itself. Essentially, I'm arguing for hierarchical modularity, rather than a flat modularity model.

> There is a case to be made for table specific crypto services.

I agree. I'm not arguing against per-table crypto. Just arguing for a slightly different interface and configuration for it, so it's easier to offload the maintenance and configuration outside of Accumulo's core responsibilities (hierarchical modularity).

**Comment by milleruntime** [comment]
> general.* would also work (and could be changed at runtime, unlike instance.*). The main point here is that these aren't characteristics of the table, but rather, characteristics of the crypto service module that inform it of how to behave in specific contexts.

Using the general prefix might work. I think being able to change the configuration at runtime is important when dealing with per-table changes. 

But I think by not including encryption as a table characteristic we are pushing more complexity to the user and creating more complexity in the code. Instead of the user configuring the service like they would other table specific classes (balancer, scan dispatcher) they have to know specialized options within the service (they will have to know some, key location etc). Then Accumulo has to deal with special circumstances of having general properties that operate on a per-table basis. How will table operations like clone or export table know how to include encryption as part of the operation? As a user, I would expect a cloned table to also be encrypted. It seems you are creating a lot of complexity just to save a few properties.

**Comment by ctubbsii** [comment]
> But I think by not including encryption as a table characteristic we are pushing more complexity to the user and creating more complexity in the code. Instead of the user configuring the service like they would other table specific classes (balancer, scan dispatcher) they have to know specialized options within the service (they will have to know some, key location etc). Then Accumulo has to deal with special circumstances of having general properties that operate on a per-table basis. How will table operations like clone or export table know how to include encryption as part of the operation? As a user, I would expect a cloned table to also be encrypted. It seems you are creating a lot of complexity just to save a few properties.

I think you're right that it makes sense to have some of this as table characteristics. I wasn't arguing against that part of the config. But your point is well-taken, and definitely leads itself to using the `table.crypto.opts.SOMETABLEOPT=SOMEVALUE` suggestion I mentioned above so that the per-table characteristics are stored with the table, rather than in the central store.

Since I was just looking at the classloader stuff a moment ago, I think I will use that as an analogy:

The classloader has a "context" property on a table. This context is a hint to the centrally managed classloader factory service. The clasloader factory is responsible for constructing a classloader, given a context. It can choose to ignore the context and provide the same classloader every time. Or, it can decide to produce a classloader that depends on the context (such as interpreting the context as a class path and returning a corresponding URLClassLoader). The classloader factory is the authority, but takes hints from the per-table characteristic.

I think this model works well, and I think it should be applied to the per-table crypto as well. I believe this is similar to how the per-table balancer works as well, as well as the per-table volume chooser. The scan dispatcher is a bit more complicated, but I think it's also similar, in that there is centralized executors in the tserver, and the per-table config (dispatcher) provides information to select an executor.

In short: centralized pluggable services, that have per-table config I think is better than having per-table services.

**Comment by PircDef** [comment]
> In short: centralized pluggable services, that have per-table config I think is better than having per-table services.

I think I must be misunderstanding something. How do you envision this for tables that are imported from users that do not use the existing service at all? I'm thinking of the word service as roughly meaning "encrypter, decrypter, and key utilities". The PR has even less, leaving the utilities up to the encrypter/decrypters.
```
public interface CryptoService {
  enum Scope {WAL, RFILE}
  FileEncrypter getEncrypter();
  FileDecrypter getDecrypter();
}
```
Lets assume that an Accumulo instance is running with the default crypto service (AESCryptoService) enabled. User A wants to import their table, and they have a custom encrypter/decrypter. Additionally, they have some custom utilities for communicating with their key server. Under this pull request, all of this custom software would be bundled under the UserACryptoService, and Accumulo would have to be configured to use this service when requested by the table. 

How do you see this working with a centralized pluggable service?

**Comment by ctubbsii** [comment]
I'm thinking something like:

```java
public interface CryptoService {
  enum Scope {WAL, RFILE}
  FileEncrypter getEncrypter(Context); // where Context contains information like {Scope=RFILE, Config=tableConfig(t1)}
  FileDecrypter getDecrypter(Context);
}
```
 
> Lets assume that an Accumulo instance is running with the default crypto service (AESCryptoService) enabled. User A wants to import their table, and they have a custom encrypter/decrypter.

I would argue for them using a CryptoService that supports their custom encrypter/decrypter features, as well as any others that the rest of the system supports. The config to indicate to the service which encrypter/decrypter features/algorithms/etc. to use can still be per-table config... but the crypto service responsible for interpreting those config options is inside the centrally configured CryptoService implementation, and not living inside Accumulo core code.

Right now, it seems that this is the workflow:

* Accumulo determines the context and gathers context information (table config, WAL config, etc.)
* Accumulo retrieves a user-configured crypto service from the context
* The crypto service provides the encrypter/decrypter for that context

My argument is for a workflow that looks like:

* Accumulo determines the context and gathers context information (table config, WAL config, etc.)
* Accumulo passes the context to the singleton crypto service
* The crypto service provides an encrypter/decrypter appropriate for the given context

This moves the decision making about which encrypter/decrypter to use into the singleton CryptoService, and removes Accumulo's responsibility to manage/instantiate multiple crypto services for different contexts, and removes the user's responsibility to configure multiple crypto services, rather than simply different crypto algorithms/options for a singleton crypto service.

**Comment by milleruntime** [comment]
Although I don't understand how the ClassLoader code works with in regards to Table context, I think I see what you are getting at. I am not sure how to implement it though so I will probably step away from this for a bit.

> Accumulo passes the context to the singleton crypto service

I thought we were trying to get away from static singleton classes? The impl that I was going for was object creation based on configuration.

**Comment by milleruntime** [comment]
OK I think it will be easier to go back to the drawing board and start a fresh branch off of what we have now in main. Closing this for now.

**Comment by ctubbsii** [comment]
> @ctubbsii did you have any ideas how to configure the decrypters? @PircDef and I both thinks we should avoid directly loading a class from bytes in a file for security reasons.

I agree.

> We thought it would be better to only load classes configured and then compare the bytes read from the file against the classes loaded for decryption.

I think you lost me a bit here. I don't see any reason we should serialize any classes to the files at all... only parameters.

For example (pseudo-code):

```java
var cryptoService = serverContext.getCryptoService();

// encrypt file
var encryptionContext = createRFileEncryptionContextForTableFromConfig({tableName, tableConfig});
var encrypter = cryptoService.getEncrypter(encryptionContext);
var encryptedOutputStream = newEncryptedOutputStream();
while (plaintext.hasMoreBytes()) {
  encryptedOutputStream.append(encrypter.encrypt(plaintext.moreBytes()));
}
var encryptedRFile = encryptedOutputStream.finish(encrypter.params());

// decrypt file
var decryptParams = encryptedRFile.readCryptoHeader(file);
var decryptionContext = createRFileDecryptionContextForTableFromFileAndConfig({tableName, tableConfig, decryptParams});
var decrypter = cryptoService.getDecrypter(decryptionContext);
var plaintextOutputStream = newPlaintextOutputStream();
while (encryptedRFile.hasMoreBytes()) {
  plaintextOutputStream.append(decrypter.decrypt(encryptedRFile.moreBytes()));
}
var plaintextRFile = plaintextOutputStream.toFile();
```

This is the basic outline, but the specific API could vary depending on where exactly the file header is read (whether it's in the decrypter itself, or inside Accumulo and merely passed to the decrypter).

Also, I believe what I just wrote out is essentially how it works today, except for the lack of having the table name and table config  as part of the encrypt/decrypt contexts. I believe that currently, the only encryption context is the opts in the system config, but the decrypt context does use the params serialized in the RFile.

**Comment by milleruntime** [comment]
> var decryptionContext = createRFileDecryptionContextForTableFromFileAndConfig({tableName, tableConfig, decryptParams});

I was wondering what the user should put in `tableConfig` for decryption. Should it be a list of classes to use for decryption? That is where I was going with this PR... User sets `table.crypto.decrypters=CryptoClassA,CryptoClassB,OldCrypto`. Then when we read the decryptParams like you have:
> var decryptParams = encryptedRFile.readCryptoHeader(file);

We look at the class name written in the `decryptParams` and see if it matches one of the classes configured in `table.crypto.decrypters`. 

Asked in another way... How do we prevent user from writing code that does bad things? Aka..
<pre>
 @Override
public FileDecrypter getFileDecrypter(CryptoEnvironment env) {
//load class bad class name
var decryptParams = env.getDecryptParams();
String maliciousClassName = decryptParams.getClassName();
return loadClass(maliciousClassName);
}

**Comment by ctubbsii** [comment]
> > var decryptionContext = createRFileDecryptionContextForTableFromFileAndConfig({tableName, tableConfig, decryptParams});
> 
> I was wondering what the user should put in `tableConfig` for decryption. Should it be a list of classes to use for decryption? That is where I was going with this PR... User sets `table.crypto.decrypters=CryptoClassA,CryptoClassB,OldCrypto`. Then when we read the decryptParams like you have:
> 
> > var decryptParams = encryptedRFile.readCryptoHeader(file);
> 
> We look at the class name written in the `decryptParams` and see if it matches one of the classes configured in `table.crypto.decrypters`.

The user shouldn't be configuring the decrypter class at all. That's too low-level to be convenient for users. The CryptoService should make that decision. Users should be configuring options that are CryptoService-dependent in `table.crypto.opts.*`. The CryptoService implementation that users configure for the system would be responsible for documenting what options are available and what they do.

> 
> Asked in another way... How do we prevent user from writing code that does bad things? Aka..
> 
>  @Override
> public FileDecrypter getFileDecrypter(CryptoEnvironment env) {
> //load class bad class name
> var decryptParams = env.getDecryptParams();
> String maliciousClassName = decryptParams.getClassName();
> return loadClass(maliciousClassName);
> }

That would be the responsibility of the CryptoService implementation itself to guard against. A good implementation of the CryptoService won't have options to specify arbitrary class names to use, but will do something more sensible... like provide options for the crypto algorithm, and it will instantiate its own decrypter implementation based on the algorithm selected.

We can't protect against users deploying a bad CryptoService implementation though. The responsibility is on them to deploy only dependencies they trust, including implementations of pluggable Accumulo components shipped in separate jars.

**Comment by milleruntime** [comment]
Another issue I ran into was how to get information down into BCFile. There is limited information for reading/writing once you get way down into the weeds:
https://github.com/apache/accumulo/blob/0781550076f04d12716650fd64881d8b9d041afa/core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/BCFile.java#L317-L331

In this PR, I created the Encrypter or List<Decrypters> based on the configuration and passed that down into BCFile. If we want to create the environment object (`encryptionContext` or `decryptionContext` in your example) down in the belly of BCFile what do we pass to it? The safest way seems to pass `AccumuloConfiguration` down but I want to avoid having to refactor all of our internal file utilities every time we want to add something to the environment object. I guess we could pass down the environment object itself (`encryptionContext` or `decryptionContext`) creating it at the higher levels. But we would still need to create the CryptoService so we can call `cryptoService.getDecrypter(decryptionContext)`
